### DISCUSSION TITLE
CA1849

### DISCUSSION THREAD

**Discussion by glen-nicol** [louthy/language-ext#1107] [Q&A]
Is there any guidance on how to elegantly avoid the CA1849 warning that occurs in any async method that uses any of the IfNone/Some/Left/Right etc functions that have Async extensions? In my case, my option doesn't require an async context to complete the IfNone result so I don't really want to pay the extra cost of allocating a task. But I also don't want to have to disable that warning globally or everywhere I use these functions with pragmas or a ton of attributes.

```csharp
public static async Task<string> Foo(Option<string> fileSuffix)
{
    var suffix = fileSuffix.IfNone(() => string.empty); // CA1849 occurs here
    var fileName = $"file{suffix}.txt";
    return await File.ReadAllTextAsync(fileName).ConfigureAwait(false);
}
```

Full warning text:
```
Warning	CA1849	'Option<string>.IfNone(Func<string>)' synchronously blocks. Await 'Option<string>.IfNoneAsync<string>(Func<string>)' instead.
When inside a Task-returning method, use the async version of methods, if they exist.
```

EDIT: changed option type to highlight I only need IfNone(). Original example required a map of int to string.

**Comment by evermanwa** [comment]
Would doing something like remedy your warning?

```csharp
var suffix = number.Match(Some: x => x.ToString(), None: () => string.Empty);
```

I couldn't reproduce it in my local environment to prompt that error, so figure I would throw this out there.

**Reply by glen-nicol** [reply]
Actually, that does avoid it. I didn't try it because I assumed it would give me the same warning because there is a MatchAsync extension too!

But this approach now requires another lambda ```x => x``` ~~```.Map(x => x)```~~ which I hope is inlined? Or atleast optimized to a static function call? I guess I have never looked too much into lambda optimization.

Also, I see my example above doesn't actually compile since I am going from int to string without a call to Map(). In my actual code there is no need for map in some cases.

**Reply by evermanwa** [reply]
In your current example, why would you need another .Map(x => x)? This in theory should satisfy your needs, although I admittedly can't read minds and am not aware of what else you could be thinking of.

**Reply by glen-nicol** [reply]
Sorry, scatterbrained. I was mixing thoughts and moved that from another sentence that I deleted without checking correct context.

it would be ```.Match(x => x, ...)``` where the extra lambda is ```x => x```

**Reply by evermanwa** [reply]
Yeah, I only explicitly put the "Some" and "None" in mine for readability. Still trying to get co-workers to embrace it, and it makes it easier to understand. You can do it that way as well.

```csharp
var suffix = number.Match(x => x.ToString(), () => string.Empty);
```

Keep in mind, that if I understand it right, both of the lamdas for the Map function will have to return that same value. One map can't return int, and the other empty string.

**Reply by glen-nicol** [reply]
Correct. I modified the original post so its all strings to clear up that confusion around not mapping the int to string. Was really trying to highlight that I just needed to cover the None case and was hoping to avoid adding an extra allocation somewhere or suppressing that warning which is helpful in other places.
